doc/architectural/background-concepts.md

```diff
@@ -764,16 +764,378 @@ Next, try running minisat on the following formula:
+```C
+unsigned long factorial(unsigned n) {
+  unsigned long fac = 1;
+  unsigned int i = 1;
+  if (i <= n) {
+    fac *= i;
+    i = i+1;
+    if (i <= n) {
+      fac *= i;
+      i = i+1;
+      if (i <= n) {
+        fac *= i;
+        i = i+1;
+        if (i <= n) {
+          IGNORE;
+        }
+    }
+  }
+  return fac;
+}
+```
+Since this program is now in a simpler form without loops, we can use the
+techniques we learned above to transform it into a propositional formula.
+First, we transform into SSA (with code guard variables already included):
+```C
+unsigned long factorial(unsigned n) {
+  unsigned long fac.1 = 1;
+  unsigned int i.1 = 1;
+  bit C1 = i.1 <= n;
+  if (C1) {
+    fac.2 = fac.1 * i.1;
+    i.2 = i.1+1;
+    bit C2 = i.2 <= n;
+    if (C2) {
+      fac.3 = fac.2 * i.2;
+      i.3 = i.2+1;
+      bit C3 = i.3 <= n;
+      if (C3) {
+        fac.4 = fac.3 * i.3;
+        i.4 = i.3+1;
+        bit C4 = i.4 <= n;
+        if (C4) {
+          IGNORE;
+        }
+    }
+  }
+  return Phi(C1, Phi(C2, Phi(C3, fac.4, fac.3), fac.2), fac.1);
+}
+```
+We translate `IGNORE` into the formula **false** - this will later allow
+us to rule out all paths that reach this point.
+
+The corresponding propositional formula can then be written as (check
+that this is equivalent to the formula you would be getting by following
+the translation procedure described above):
+
+> fac.1 = 1 and i.1 = 1 and C1 = i.1 &lt;= n and
+> ((not C1) &rArr; return = fac.1) and
+> C1 &rArr; (
+>>   fac.2 = fac.1 * i.1 and i.2 = i.1 + 1 and C2 = i.2 &lt;= n and
+>>   ((not C2) &rArr; return = fac.2) and
+>>   C2 &rArr; (
+>>>     fac.3 = fac.2 * i.2 and i.3 = i.2 + 1 and C3 = i.3 &lt;= n and
+>>>     ((not C3) &rArr; return = fac.3) and
+>>>     C3 &rArr; (
+>>>>       fac.4 = fac.3 * i.3 and i.4 = i.3 + 1 and C4 = i.4 &lt;= n and
+>>>>       ((not C4) &rArr; return = fac.4) and
+>>>>       (C4 &rArr; false)
+>>>     )
+>>>   )
+>> )
+
+In the following, we reference this formula as FA(n, result).
+
+At this point, we know how to encode programs as propositional formulas.
+Our goal was to reason about programs, and in particular, to check whether
+a certain property holds. Suppose, for example, that we want to check if there
+is a way that the `factorial` function returns `6`. One way to do this is to
+look at the following propositional formula: FA(n, result) and result = 6.
+If this formula has a model (i.e., if we can find a satisfying assignment to
+all variables, and in particular, to n), we can extract the required value
+for the parameter `n` from that model. As we have discussed above, this can
+be done using a SAT solver: If you run, say, MiniSAT on this formula, you will
+get a model involving n=3.
+
+Be aware that this method has very clear limits: We know that the factorial of
+`5` is `120`, but with the formula above, evaluating
+''FA(n, result) and result=120'' would yield ''unsatisfiable''! This is because
```

Comment:
Maybe state explicitly that 'VERIFICATION SUCCESSFUL' is to be interpreted modulo the bounds given.

Reply:
Added